### PR TITLE
Make pdf work

### PR DESCRIPTION
**Pull request by vicancy**
This version keeps OPS logic that:
1. PDF is generated from TOC file, each TOC file generates a PDF file
2. Not-in-TOC files are collected into a appendices PDF file

### CODE REVIEW

**Comment by vwxyzh** [comment]
Full HTML:
```
<html xmlns="http://www.w3.org/1999/xhtml"><head></head><body></body></html>
```

File: src/docfx/SubCommands/PdfCommand.cs [hunk 1]
```diff
@@ -4,25 +4,130 @@
...
+                PdfConvertParallelism = _config.MaxParallelism == null || _config.MaxParallelism <= 0 ? Environment.ProcessorCount : _config.MaxParallelism.Value,
+                PdfDocsetName = _config.Name ?? Path.GetFileName(EnvironmentContext.BaseDirectory),
+                SourceDirectory = outputFolder,
+                ExcludeTocs = _config.ExcludedTocs?.ToArray(),
+            };
+
+            // 1. call BuildCommand to generate html files first
+            _innerBuildCommand.Exec(context);
+            // 2. call html2pdf converter
+
```
Comment (superyyrrzz) [inline src/docfx/SubCommands/PdfCommand.cs]: minor: move empty line above

File: src/docfx.website.themes/pdf.default/layout/_master.tmpl [hunk 2]
```diff
@@ -0,0 +1,11 @@
+{{!Copyright (c) Microsoft. All rights reserved. Licensed under the MIT license. See LICENSE file in the project root for full license information.}}
+{{!include(/^styles/.*/)}}
+<!DOCTYPE html>
+<!--[if IE]><![endif]-->
+<html>
+  <body>
+    <article>
+        {{!body}}
```
Comment (superyyrrzz) [inline src/docfx.website.themes/pdf.default/layout/_master.tmpl]: minor: indent -2?

File: src/Microsoft.DocAsCode.HtmlToPdf/HtmlToPdfConverter.cs [hunk 3]
```diff
@@ -13,6 +13,7 @@ namespace Microsoft.DocAsCode.HtmlToPdf
     using System.Web;
 
     using iTextSharp.text.pdf;
+    using Microsoft.DocAsCode.Plugins;
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.HtmlToPdf/HtmlToPdfConverter.cs]: group

**Comment by hellosnow** [comment]
>                     if (options.Content == null && options.Resource == null)

the logic to check options could be moved into                         MergeOptionsToConfig? if the config file not exist, return null directly?

---
Refers to: src/docfx/SubCommands/BuildCommand.cs:113 in c73e5f3. [](commit_id = c73e5f35f03e123667d9dc042f159f04acdd3ec3, deletion_comment = False)

File: src/docfx/Models/PdfCommandOptions.cs [hunk 4]
```diff
@@ -11,16 +11,25 @@ internal class PdfCommandOptions : BuildCommandOptions
         [Option("name", HelpText = "Specify the name of the generated pdf")]
         public string Name { get; set; }
 
+        [Option("css", HelpText = "Specify the path for the css to generate pdf, default value is styles/default.css")]
+        public string CssFilePath { get; set; }
+
         [Option("appendices", HelpText = "Specify whether or not generate appendices for not-in-TOC articles")]
         public bool GenerateAppendices { get; set; }
 
         [Option("external", HelpText = "Specify whether or not generate external links for PDF")]
```
Comment (hellosnow) [inline src/docfx/Models/PdfCommandOptions.cs]: >external [](start = 17, length = 8)

minor: better option name for bool value, for example "enableExternalLink"?

File: src/Microsoft.DocAsCode.HtmlToPdf/ConvertWrapper.cs [hunk 5]
```diff
@@ -209,24 +213,16 @@ private void AbsolutePathInTocPageFileTransformer(string tocPage)
...
-            return LoadFromFilePath<IList<TocModel>>(Path.Combine(basePath, tocFile.Output.TocJson.RelativePath));
+            return JsonUtility.Deserialize<IList<TocModel>>(Path.Combine(basePath, ManifestUtility.GetRelativePath(tocFile, OutputType.TocJson)));
         }
 
         private IEnumerable<string> GetManifestHtmls(Manifest manifest)
         {
             return from file in manifest.Files
-                   where file != null && file.Type == ManifestItemType.Content && file.Output != null && file.Output.Html != null
-                   let outputPath = file.Output.Html.RelativePath
+                   where file != null && IsType( file.DocumentType, ManifestItemType.Content)
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.HtmlToPdf/ConvertWrapper.cs]: >  [](start = 48, length = 1)

minor: extra space...

File: src/Microsoft.DocAsCode.HtmlToPdf/ConvertWrapper.cs [hunk 6]
```diff
@@ -282,6 +278,21 @@ private void ConvertCore(string basePath, string pdfFileName, IList<HtmlModel> h
 
             converter.Save(Path.Combine(_pdfOptions.DestDirectory, pdfFileName));
         }
+        private bool IsType(string type, ManifestItemType targetType)
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.HtmlToPdf/ConvertWrapper.cs]: >private [](start = 8, length = 7)

minor: prepend one line

File: src/Microsoft.DocAsCode.HtmlToPdf/ConvertWrapper.cs [hunk 7]
```diff
@@ -282,6 +278,21 @@ private void ConvertCore(string basePath, string pdfFileName, IList<HtmlModel> h
...
+        {
+            if (Enum.TryParse(type, out ManifestItemType actualType))
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.HtmlToPdf/ConvertWrapper.cs]: >TryParse [](start = 21, length = 8)

ignore case? and if can parse from enum, still need to compare with target type?

**Comment by hellosnow** [comment]
:shipit:

**Comment by vicancy** [comment]
which one do you mention?

---
In reply to: [300960911](https://github.com/dotnet/docfx/pull/1658#issuecomment-300960911) [](ancestors = 300960911)

**Pull request merged**